File: mobile/android/core/designsystem/src/main/kotlin/com/slopssaloon/omen/core/designsystem/token/OmenFocusRing.kt
package com.slopssaloon.omen.core.designsystem.token

import androidx.compose.foundation.border
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Semantic focus/selection outline (registry §4; m1-focus-ring-build-brief-v1.md).
 *
 * Draws a two-layer outline — a soft halo plus a crisp stroke — derived from
 * [OmenColorScheme.focusRing] / [OmenColorScheme.focusRingHalo]. `Modifier.border` draws inside
 * the existing layout bounds, so this never changes a component's measured size.
 *
 * This modifier supplies the *visible outline* half of the non-color focus contract only.
 * Callers remain responsible for the other half: real Compose focus/interaction state
 * ([androidx.compose.foundation.interaction.InteractionSource] / native `focusable()`) and, for
 * selected controls, an additional shape/weight/checkmark cue — color and outline alone are
 * never sufficient (registry §4, brief §2/§5).
 *
 * @param focused whether the owning component currently holds focus or accessibility focus.
 * @param color the ring stroke color, normally [OmenColorScheme.focusRing].
 * @param haloColor the soft underlay color, normally [OmenColorScheme.focusRingHalo].
 * @param shape corner shape to match the target component; defaults to a 8dp rounded rect.
 */
fun Modifier.omenFocusRing(
    focused: Boolean,
    color: Color,
    haloColor: Color,
    shape: RoundedCornerShape = RoundedCornerShape(8.dp),
    strokeWidth: Dp = 2.dp,
    haloWidth: Dp = 4.dp,
): Modifier = if (!focused) {
    this
} else {
    this
        .border(width = haloWidth, color = haloColor, shape = shape)
        .border(width = strokeWidth, color = color, shape = shape)
}

File: mobile/ios/OmenIOS/OmenIOS/DesignSystem/OmenFocusRing.swift
import SwiftUI

/// Semantic focus/selection outline (registry §4; m1-focus-ring-build-brief-v1.md).
///
/// Draws a two-layer outline — a soft halo plus a crisp stroke — derived from `OmenColor`.
/// Applied via `.overlay`, so it never changes a component's measured size (brief §3, "Adds a
/// visible outline without changing layout size").
///
/// This modifier supplies the *visible outline* half of the non-color focus contract only.
/// Callers remain responsible for the other half: real SwiftUI focus state (`@FocusState`) or
/// accessibility-focus, and, for selected controls, an additional shape/weight/checkmark cue —
/// color and outline alone are never sufficient (registry §4, brief §2/§5).
struct OmenFocusRing: ViewModifier {
    let focused: Bool
    var color: Color = OmenColor.focusRing
    var haloColor: Color = OmenColor.focusRingHalo
    var cornerRadius: CGFloat = 8
    var strokeWidth: CGFloat = 2
    var haloWidth: CGFloat = 4

    func body(content: Content) -> some View {
        content.overlay(
            ZStack {
                if focused {
                    RoundedRectangle(cornerRadius: cornerRadius)
                        .stroke(haloColor, lineWidth: haloWidth)
                    RoundedRectangle(cornerRadius: cornerRadius)
                        .stroke(color, lineWidth: strokeWidth)
                }
            }
        )
    }
}

extension View {
    /// - Parameter isFocused: whether the owning control currently holds focus or accessibility
    ///   focus. Pass a `@FocusState` value or an equivalent native focus/selection source —
    ///   never a synthesized app-level focus model (brief §3).
    func omenFocusRing(
        isFocused: Bool,
        color: Color = OmenColor.focusRing,
        haloColor: Color = OmenColor.focusRingHalo,
        cornerRadius: CGFloat = 8
    ) -> some View {
        modifier(
            OmenFocusRing(
                focused: isFocused,
                color: color,
                haloColor: haloColor,
                cornerRadius: cornerRadius
            )
        )
    }
}
